### PR TITLE
xds: handle errors in cds balancer

### PR DESCRIPTION
**Pull request by menghanl**
This is similar to the EDS error handling: https://github.com/grpc/grpc-go/pull/3645

### CODE REVIEW

File: xds/internal/balancer/cdsbalancer/cdsbalancer.go [hunk 1]
```diff
@@ -271,6 +265,45 @@ func (b *cdsBalancer) run() {
...
+func (b *cdsBalancer) handleErrorFromUpdate(err error, fromParent bool) {
+	// TODO: connection errors will be sent to the eds balancers directly, and
+	// also forwarded by the parent balancers/resolvers. So the eds balancer may
+	// see the same error multiple times. We way want to only forward the error
+	// to eds if it's not a connection error.
+	//
+	// This is not necessary today, because xds client never sends connection
+	// errors.
+
+	if fromParent && xdsclient.ErrType(err) == xdsclient.ErrorTypeResourceNotFound && b.cancelWatch != nil {
```
Comment (easwars) [inline xds/internal/balancer/cdsbalancer/cdsbalancer.go:292]: Optional: Might be a bit more readable to have the check for `cancelWatch` inside the top if.

File: xds/internal/balancer/cdsbalancer/cdsbalancer_test.go [hunk 2]
```diff
@@ -398,36 +407,127 @@ func (s) TestHandleClusterUpdate(t *testing.T) {
...
+	xdsC.InvokeWatchClusterCallback(xdsclient.ClusterUpdate{}, err1)
+	if err := xdsC.WaitForCancelClusterWatch(); err == nil {
+		t.Fatal("watch was canceled, want not canceled (timeout error)")
+	}
+	if err := edsB.waitForResolverError(err1); err == nil {
+		t.Fatal("eds balancer shouldn't get error (shouldn't be built yet)")
+	}
+	if state, err := tcc.newPickerCh.Receive(); err != nil {
+		t.Fatalf("failed to get picker, expect an error picker")
+	} else {
```
Comment (easwars) [inline xds/internal/balancer/cdsbalancer/cdsbalancer_test.go:434]: We don't need this else. The code in the else block can be unconditional.

File: xds/internal/balancer/cdsbalancer/cdsbalancer.go (hunk 1 above)
Comment (menghanl) [reply xds/internal/balancer/cdsbalancer/cdsbalancer.go]: Done

File: xds/internal/balancer/cdsbalancer/cdsbalancer_test.go (hunk 2 above)
Comment (menghanl) [reply xds/internal/balancer/cdsbalancer/cdsbalancer_test.go]: We need else to capture the `state` from `if`.
Or, I move `state, err := tcc.newPickerCh.Receive()` out from `if`.

Comment (easwars) [reply xds/internal/balancer/cdsbalancer/cdsbalancer_test.go]: Makes sense. I missed that.

**Pull request merged**